### PR TITLE
Do not apply Stripe application fee if Stripe account country is Brazil

### LINKED ISSUE #5729: Stripe fee should not apply for Brazilian Stripe accounts.
## Details

Instead of ignoring the Stripe application fee, Stripe returns errors and which prevent donation if Stripe account is from Brazil country.

Stripe says,
```
Stripe is not able to support application fees with connected accounts in Brazil. For this reason, charges created by these connected accounts will receive an error code. In order to keep your platform working in Brazil, you need to set your application fees to zero for all connected accounts in Brazil (in your PaymentIntent API integration) as an immediate solution and arrange with the connected accounts alternative ways to collect the fees related to your services.
```

## Acceptance Criteria

-   [x] Stripe fee does apply for United States (US) Stripe accounts.
-   [x] Stripe fee does not apply for Brazilian (BR) Stripe accounts.

### PR DESCRIPTION
**Pull request by ravinderk**
Resolves #5729 

## Description

Added logic to prevent application fee for Brazilian Stripe account.

## Testing Instructions

Follow these [testing instructions](https://github.com/impress-org/givewp/pull/5709). You can also use the Brazilian Stripe test account (if you have).

## Pre-review Checklist

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@since` tags included in DocBlocks
-   [x] Changes logged to the `Unreleased` section of `CHANGELOG.md`
-   [x] Includes unit and/or end-to-end tests
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

File: src/PaymentGateways/Stripe/ApplicationFee.php [hunk 1]
```diff
@@ -28,14 +29,21 @@ class ApplicationFee {
...
 		$gate = new static();
 
-		return ! ( $gate->hasLicense()
-			|| $gate->isStripeProAddonActive()
-			|| $gate->isStripeProAddonInstalled( get_plugins() ) );
+		/* @var $accountDetailRepository AccountDetail */
+		$accountDetailRepository = give( AccountDetail::class );
+		$stripeAccountDetail     = $accountDetailRepository->getAccountDetail( $accountId );
+
+		return $gate->isStripeProAddonActive()
```
Comment (kjohnson) [inline src/PaymentGateways/Stripe/ApplicationFee.php:43]: @ravinderk I think this reverses the logic of some of the gates from the previous commit - could you review this part?

Comment (ravinderk) [reply src/PaymentGateways/Stripe/ApplicationFee.php:43]: @kjohnson you are right. I will fix it.

File: src/PaymentGateways/Stripe/Models/AccountDetail.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	 * @unreleased
+	 * @param array $args
+	 *
+	 * @return AccountDetail
+	 */
+	public static function fromArray( $args ) {
+		$accountDetail = new static();
+		$accountDetail->validate( $args );
+
+		$accountDetail->type               = $args['type'];
```
Comment (kjohnson) [inline src/PaymentGateways/Stripe/Models/AccountDetail.php:37]: I think we can reduce the overhead of this class by extending an [ArrayObject](https://www.php.net/manual/en/class.arrayobject.php), in which case the only definition required is the list of required keys used by `validate`.

File: src/PaymentGateways/Stripe/ApplicationFee.php (hunk 1 above)
Comment (ravinderk) [reply src/PaymentGateways/Stripe/ApplicationFee.php]: @kjohnson condition has been fixed

File: src/PaymentGateways/Stripe/Repositories/AccountDetail.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	 * Return Stripe account id for donation form.
+	 *
+	 * @unreleased
+	 * @param int $formId
+	 *
+	 * @return \Give\PaymentGateways\Stripe\Models\AccountDetail
+	 */
+	public function getDonationFormStripeAccountId( $formId ) {
+		// Global Stripe account.
+		$accountId = give_get_option( '_give_stripe_default_account', '' );
```
Comment (kjohnson) [inline src/PaymentGateways/Stripe/Repositories/AccountDetail.php:22]: I would suggest moving the `$accontId` default closer to its use. For Example, it might fit better as an `else` to the `if` check. Otherwise it is easy to miss the intent if a form doesn't not have Stripe enabled.

File: tests/unit/tests/PaymentGateways/Stripe/ApplicationFeeTest.php [hunk 4]
```diff
@@ -91,4 +91,18 @@ public function testHasLicenseAllAccessPass() {
             $this->gate->hasLicense( 'give-stripe' )
         );
     }
+
+    public function testIsStripeAccountFromBrazil() {
```
Comment (kjohnson) [inline tests/unit/tests/PaymentGateways/Stripe/ApplicationFeeTest.php:95]: Let's rephrase this as a test of `isCountrySupportApplicationFee` as opposed to the specific instance of Brazil.

File: src/PaymentGateways/Stripe/ApplicationFee.php [hunk 5]
```diff
@@ -28,14 +29,19 @@ class ApplicationFee {
 	 *
 	 * @unreleased
 	 *
+	 * @param string $accountId
+	 *
 	 * @return bool
 	 */
-	public static function canAddFee() {
+	public static function canAddFee( $accountId ) {
```
Comment (kjohnson) [inline src/PaymentGateways/Stripe/ApplicationFee.php:36]: Instead of threading `$accountId` through `canAddFee`, can we move this logic closer to the `isCountrySupportApplicationFee` check itself? It is the only check that needs it.

File: src/PaymentGateways/Stripe/Models/AccountDetail.php (hunk 2 above)
Comment (ravinderk) [reply src/PaymentGateways/Stripe/Models/AccountDetail.php:37]: @kjohnson Can you show me how? I did not understand you.

File: src/PaymentGateways/Stripe/ApplicationFee.php (hunk 5 above)
Comment (ravinderk) [reply src/PaymentGateways/Stripe/ApplicationFee.php:36]: @Kyle B. Johnson can you suggest something?

File: src/PaymentGateways/Stripe/Models/AccountDetail.php (hunk 2 above)
Comment (kjohnson) [reply src/PaymentGateways/Stripe/Models/AccountDetail.php:37]: Actually, since we aren't iterating on it, maybe  not ArrayObject.

But, it might make sense to use the magic method `__get()` to map properties to array keys.

```php
class AccountDetail {
  protected $args;
  
  public function __construct( $args ) {
    $this->args = $args;
  }
  
  public function __get( $key ) {
    return $this->args[ $key ];
  }
}
```

Comment (kjohnson) [reply src/PaymentGateways/Stripe/Models/AccountDetail.php:37]: Expanded to include validation.

```php
class AccountDetail {
  protected $args;
  protected $requiredArgs;
  
  public function __construct( $args ) {
    $this->args = $args;
    $this->validate();
  }
  
  public function __get( $key ) {
    return $this->args[ $key ];
  }

  public function validate() {
    // Compare $args to $requiredArgs
  }
}
```

File: src/PaymentGateways/Stripe/ApplicationFee.php (hunk 5 above)
Comment (kjohnson) [reply src/PaymentGateways/Stripe/ApplicationFee.php:36]: Since `give_stripe_get_connected_account_options` is pulling off of the session, can we call it directly from the check instead of passing through the `$accountId`?

Comment (kjohnson) [reply src/PaymentGateways/Stripe/ApplicationFee.php:36]: Then we could drop the changes to the `includes` files and the change would be contained to the new code.

Comment (ravinderk) [reply src/PaymentGateways/Stripe/ApplicationFee.php:36]: @kjohnson I will not suggest that because that way this function output depends upon the donation section and we of this function also limit to donation processing context.

Comment (kjohnson) [reply src/PaymentGateways/Stripe/ApplicationFee.php:36]: @ravinderk what if we use the container to resolve a `Modle\AccountDetail` from the repository and inject the `give_stripe_get_connected_account_options()` there:

```php
// Service Provider
$this->bind( AccountDetail::class, function() {
    $accountOptions = give_stripe_get_connected_account_options();
    return ( new Repository\AccountDetail( ) )->getAccouuntDetail( $account_options['stripe_account'] );
});

```

Comment (ravinderk) [reply src/PaymentGateways/Stripe/ApplicationFee.php]: @kbjohnson90 I feel that we need a class that returns a result of ApplicationFee in a donation processing context. I added a new singleton class. In this way, we can use the `ApplicationFee` independently.

commit: https://github.com/impress-org/givewp/compare/f3ebb6dcb2ff...ac859d6fa180

```php
class ProcessingDonation {
	/**
	 * @var Models\AccountDetail
	 */
	protected $accountDetail;

	public static function canAddFee() {
		/* @var $donationProcessing self */
		$donationProcessing = give( __CLASS__ );
		$gate               = new ApplicationFee( $donationProcessing->accountDetail );

		return $gate->canAddFee();
	}
	
	public function setAccountDetail() {
		// code
		return $this;
	}
}


// in service provider boot function.
give()->singleton( ProcessingDonation::class,
	function() {
		return ( new ProcessingDonation() )->setAccountDetail();
	}
);
```

Comment (kjohnson) [reply src/PaymentGateways/Stripe/ApplicationFee.php]: @ravinderk I like where you are going with this. Would it make sense to inject the account details directly into the ApplicationFee class via the container? I'm not sure if the additional class is necessary - I think we can combine these two as they are really doing one thing.

Comment (ravinderk) [reply src/PaymentGateways/Stripe/ApplicationFee.php]: @kbjohnson90 I am loading `AccuntDetail` model into `ApplicationFee` class. it looks like:

```php
class ApplicationFee {
	//code
	
	/**
	 * ApplicationFee constructor.
	 *
	 * @param  Models\AccountDetail  $accountDetail
	 */
	public function __construct( Models\AccountDetail $accountDetail ) {
		$this->accountDetail = $accountDetail;
	}
	
	//code
}
``` 

 `ProcessingDonation` and `ApplicationFee` class seems similar but they are.  `ProcessingDonation` bound to context while the `ApplicationFee` class can be used independently. For some reason `ProcessingDonation::canAddFee` is easily accessible then `ApplicationFee`.
 
 ```php
 // use 1
if ( ProcessingDonation::canAddFee() ) {
    // code
}
 
 // use 2
$accountDetailRepository = give( AccountDetail::class );
$accountId  = give_stripe_get_connected_account_options()['stripe_account'];
$applicationFee = new ApplicationFee( $accountDetailRepository->getAccountDetail( $accountId ) );

if( applicationFee->canAddFee() ){
    // code
}
 
 ```

**Comment by kjohnson** [comment]
@ravinderk I think at this point we just need to carefully test the change, as this affects the entire Application Fee feature.

**Comment by JasonTheAdams** [comment]
Please do not commit this today for the 2.10.0 release. We will get out another release in a couple weeks after further testing.

**Comment by kjohnson** [comment]
@JasonTheAdams agreed!

**Review by kjohnson — COMMENTED** [review]
@ravinderk sorry to take so long to review this. I keep coming back to it and I can't help but get the feeling that we have introduced too many levels of abstraction for what could otherwise be a simple procedural check. The `canAddFee()` method is already a wrapper for a set of procedural checks and I think it might make sense to simply add another procedural check instead of refactoring the implementing system.

**Comment by kjohnson** [comment]
@ravinderk here are my suggestions, see https://github.com/impress-org/givewp/pull/5784/commits/cf95f758323e8bdc47865cb8bfd6c3619c637e4f and https://github.com/impress-org/givewp/pull/5784.

I really like where you are going with your object oriented approach and you have shown a lot of growth in how you think about solutions.

My suggestions are basically just to simplify the overhead of the solution.

**Comment by ravinderk** [comment]
> @ravinderk here are my suggestions, see cf95f75 and #5784.
> 
> I really like where you are going with your object oriented approach and you have shown a lot of growth in how you think about solutions.
> 
> My suggestions are basically just to simplify the overhead of the solution.

The suggestion has been implemented. Can you review this pull request?

**Review by kjohnson — APPROVED** [review]
